test/mb_debug_optimus.py

```diff
@@ -0,0 +1,667 @@
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+
+    op_flag_summary_stats.process_single(op_flags, primary_only=False)
+    cr_flag_summary_stats.process_single(cr_flags, primary_only=False)
+# EPY: END code
+
+# EPY: START code
+op_primary_only_flag_summary_stats = FlagSummaryStatistics()
+cr_primary_only_flag_summary_stats = FlagSummaryStatistics()
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+
+    op_primary_only_flag_summary_stats.process_single(op_flags, primary_only=True)
+    cr_primary_only_flag_summary_stats.process_single(cr_flags, primary_only=True)
+# EPY: END code
+
+# EPY: START code
+primary_only_flag_xor_summary_stats = FlagSummaryStatistics()
+primary_only_flag_and_summary_stats = FlagSummaryStatistics()
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+
+    primary_only_flag_xor_summary_stats.process_difference(
+        op_flags, cr_flags, primary_only=True, joint_op='xor')
+
+    primary_only_flag_and_summary_stats.process_difference(
+        op_flags, cr_flags, primary_only=True, joint_op='and')
+# EPY: END code
+
+# EPY: START code
+print(f'Number of Optimus reads with multiple primary alignments: {op_multiple_primary}, fraction = {op_multiple_primary/num_total_records:.4f}')
+print(f'Number of CellRanger reads with multiple primary alignments: {cr_multiple_primary}, fraction = {cr_multiple_primary/num_total_records:.4f}')
+print()
+
+print('Optimus (primary only):')
+op_primary_only_flag_summary_stats.print_summary()
+print()
+
+print('CellRanger (primary only):')
+cr_primary_only_flag_summary_stats.print_summary()
+print()
+
+print('joint XOR (primary only):')
+primary_only_flag_xor_summary_stats.print_summary()
+print()
+
+print('joint AND (primary only):')
+primary_only_flag_and_summary_stats.print_summary()
+print()
+
+print('Optimus (all):')
+op_flag_summary_stats.print_summary()
+print()
+
+print('CellRanger (all):')
+cr_flag_summary_stats.print_summary()
+print()
+# EPY: END code
+
+# EPY: START markdown
+# **Summary**
+# 
+# - There are about ~ 70k identical raw reads in the synthetic data and they need to be removed
+# - For primary alignments with discordant flags, 47850 reads are marked as duplicate by both Optimus and CellRanger
+#   whereas 446777 reads (about half of all reads) are marked duplicate differently
+# - CellRanger marks more primary alignments as duplicate (356877 vs 185600), however, this does not necessary
+#   imply more molecules.
+# - It looks like `sctools` does not use the duplicate flag at all; `MarkDuplicates` only _marks_ duplicates. It does not remove them.
```

Comment:
That's true, but it _should_ use the duplicates flag. It'd be a faster way to construct the count matrix. I think we could just add up all the non-duplicate reads.